[Mappings editor] Flatten field type list

alisonelizabeth opened the pull request:
This PR removes the `specialised`, `geo` and `date` subtypes  and moves the values to the main field type field. It also changes the select field to a combo box. I left `range` and `numeric` sub types.

![i9vWDfQHbo](https://user-images.githubusercontent.com/5226211/70339119-4096d880-181c-11ea-8628-acc18691e9ce.gif)

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13955/)
* Commit: 757d39635873e5a4a351b5044772598f1c3f2f64

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/)
* Commit: e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/","number":"14053","commit":"e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf"}]}
PIPELINE-->

alisonelizabeth commented:
This PR is ready for review!

I added a documentation link as the first pass at providing more info to the user about each type. I'm open to changing the text or maybe even changing to an info icon instead? 

<img width="1268" alt="Screen Shot 2019-12-09 at 10 27 00 AM" src="https://user-images.githubusercontent.com/5226211/70449014-3e788800-1a6f-11ea-80fe-5a29dc77e6ad.png">

<img width="1246" alt="Screen Shot 2019-12-09 at 10 27 16 AM" src="https://user-images.githubusercontent.com/5226211/70449015-3e788800-1a6f-11ea-8de7-27fe52e6a539.png">

The `ComboBox` does provide a `renderOption` prop to customize the options. However, it requires each option have the same height, so we can't quite achieve something like the `SuperSelect`. Design also brought up a good point, that adding descriptions per option could be overwhelming if there are many options.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/)
* Commit: 7f3cb616afe8c3ff0f4820a8823ab65884eeaa00

### History
* :broken_heart: [Build #14053](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/) failed e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/","number":"14053","commit":"e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/","number":"14220","commit":"7f3cb616afe8c3ff0f4820a8823ab65884eeaa00"}]}
PIPELINE-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/)
* Commit: b0d97eac13d732cbc5be2fa2576743ecd4e094df

### History
* :broken_heart: [Build #14220](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/) failed 7f3cb616afe8c3ff0f4820a8823ab65884eeaa00
* :broken_heart: [Build #14053](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/) failed e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/","number":"14053","commit":"e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/","number":"14220","commit":"7f3cb616afe8c3ff0f4820a8823ab65884eeaa00"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/","number":"14247","commit":"b0d97eac13d732cbc5be2fa2576743ecd4e094df"}]}
PIPELINE-->

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/field_parameters/type_parameter.tsx:
```diff
@@ -0,0 +1,62 @@
+          label={typeField.label}
+          error={error}
+          isInvalid={isInvalid}
+          labelAppend={
```
@cjcenizal great idea! Made the change. Let me know what you think.

<img width="1233" alt="Screen Shot 2019-12-10 at 2 56 34 PM" src="https://user-images.githubusercontent.com/5226211/70564113-8d045000-1b5d-11ea-8cf2-ba4ff5033a13.png">
<img width="1237" alt="Screen Shot 2019-12-10 at 2 56 53 PM" src="https://user-images.githubusercontent.com/5226211/70564114-8d045000-1b5d-11ea-953b-c9cf214c1502.png">

cjcenizal commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/fields/create_field/create_field.tsx:
```diff
@@ -124,59 +126,93 @@ export const CreateField = React.memo(function CreateFieldComponent({
       const { subTypeOptions, subTypeLabel } = getSubTypeMeta(type);
 
+      const docLink =
+        (documentationService.getTypeDocLink(subType) as string) ||
```
Range and Numeric are the only data types that still have sub-types, right? If so, then maybe we should just keep the help text under the "Type" combo box, since there are only Range and Numeric help pages, and no help pages for the individual sub-types.

https://www.elastic.co/guide/en/elasticsearch/reference/master/number.html
https://www.elastic.co/guide/en/elasticsearch/reference/master/range.html

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/)
* Commit: 0907a5fc4f29136653b6579f5d74ef8ae6a76e80

### History
* :green_heart: [Build #14247](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/) succeeded b0d97eac13d732cbc5be2fa2576743ecd4e094df
* :broken_heart: [Build #14220](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/) failed 7f3cb616afe8c3ff0f4820a8823ab65884eeaa00
* :broken_heart: [Build #14053](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/) failed e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/","number":"14053","commit":"e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/","number":"14220","commit":"7f3cb616afe8c3ff0f4820a8823ab65884eeaa00"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/","number":"14247","commit":"b0d97eac13d732cbc5be2fa2576743ecd4e094df"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/","number":"14630","commit":"0907a5fc4f29136653b6579f5d74ef8ae6a76e80"}]}
PIPELINE-->

alisonelizabeth commented:
@cjcenizal I updated the code so that the doc link only displays under the main field type.

I also found a bug. With the ComboBox component, you can select an option with the enter key. This was also causing the form to submit. I don't think this is desired behavior, especially for fields like `alias` that have required parameters. I updated the code to prevent this behavior via https://github.com/elastic/kibana/pull/52400/commits/9da4267c9c564724fe01bc5f960d2b14c1501374.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/)
* Commit: 7423f34b951a068d86c7d00f8ad7bf688c1a6b2b

### History
* :green_heart: [Build #14630](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/) succeeded 0907a5fc4f29136653b6579f5d74ef8ae6a76e80
* :green_heart: [Build #14247](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/) succeeded b0d97eac13d732cbc5be2fa2576743ecd4e094df
* :broken_heart: [Build #14220](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/) failed 7f3cb616afe8c3ff0f4820a8823ab65884eeaa00
* :broken_heart: [Build #14053](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/) failed e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/","number":"14053","commit":"e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/","number":"14220","commit":"7f3cb616afe8c3ff0f4820a8823ab65884eeaa00"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/","number":"14247","commit":"b0d97eac13d732cbc5be2fa2576743ecd4e094df"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/","number":"14630","commit":"0907a5fc4f29136653b6579f5d74ef8ae6a76e80"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/","number":"14836","commit":"7423f34b951a068d86c7d00f8ad7bf688c1a6b2b"}]}
PIPELINE-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/)
* Commit: 266163c73405518a509d38ae84287934bddd8bd2

### History
* :green_heart: [Build #14836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/) succeeded 7423f34b951a068d86c7d00f8ad7bf688c1a6b2b
* :green_heart: [Build #14630](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/) succeeded 0907a5fc4f29136653b6579f5d74ef8ae6a76e80
* :green_heart: [Build #14247](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/) succeeded b0d97eac13d732cbc5be2fa2576743ecd4e094df
* :broken_heart: [Build #14220](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/) failed 7f3cb616afe8c3ff0f4820a8823ab65884eeaa00
* :broken_heart: [Build #14053](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/) failed e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14053/","number":"14053","commit":"e8b6be3ccf2bb7974c477e1bd524c9016ae1d5bf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/","number":"14220","commit":"7f3cb616afe8c3ff0f4820a8823ab65884eeaa00"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/","number":"14247","commit":"b0d97eac13d732cbc5be2fa2576743ecd4e094df"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/","number":"14630","commit":"0907a5fc4f29136653b6579f5d74ef8ae6a76e80"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/","number":"14836","commit":"7423f34b951a068d86c7d00f8ad7bf688c1a6b2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/","number":"14855","commit":"266163c73405518a509d38ae84287934bddd8bd2"}]}
PIPELINE-->

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14873/)
* Commit: ac47fc5586720077be30c5f09116a36a7d9372fe

### History
* :green_heart: [Build #14855](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/) succeeded 266163c73405518a509d38ae84287934bddd8bd2
* :green_heart: [Build #14836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/) succeeded 7423f34b951a068d86c7d00f8ad7bf688c1a6b2b
* :green_heart: [Build #14630](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/) succeeded 0907a5fc4f29136653b6579f5d74ef8ae6a76e80
* :green_heart: [Build #14247](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/) succeeded b0d97eac13d732cbc5be2fa2576743ecd4e094df
* :broken_heart: [Build #14220](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/) failed 7f3cb616afe8c3ff0f4820a8823ab65884eeaa00

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14220/","number":"14220","commit":"7f3cb616afe8c3ff0f4820a8823ab65884eeaa00"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/","number":"14247","commit":"b0d97eac13d732cbc5be2fa2576743ecd4e094df"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/","number":"14630","commit":"0907a5fc4f29136653b6579f5d74ef8ae6a76e80"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/","number":"14836","commit":"7423f34b951a068d86c7d00f8ad7bf688c1a6b2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/","number":"14855","commit":"266163c73405518a509d38ae84287934bddd8bd2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14873/","number":"14873","commit":"ac47fc5586720077be30c5f09116a36a7d9372fe"}]}
PIPELINE-->

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15139/)
* Commit: f64cd3fee556435843fd61e55d57794ef841e79d

### History
* :broken_heart: [Build #14873](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14873/) failed ac47fc5586720077be30c5f09116a36a7d9372fe
* :green_heart: [Build #14855](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/) succeeded 266163c73405518a509d38ae84287934bddd8bd2
* :green_heart: [Build #14836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/) succeeded 7423f34b951a068d86c7d00f8ad7bf688c1a6b2b
* :green_heart: [Build #14630](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/) succeeded 0907a5fc4f29136653b6579f5d74ef8ae6a76e80
* :green_heart: [Build #14247](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/) succeeded b0d97eac13d732cbc5be2fa2576743ecd4e094df

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14247/","number":"14247","commit":"b0d97eac13d732cbc5be2fa2576743ecd4e094df"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/","number":"14630","commit":"0907a5fc4f29136653b6579f5d74ef8ae6a76e80"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/","number":"14836","commit":"7423f34b951a068d86c7d00f8ad7bf688c1a6b2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/","number":"14855","commit":"266163c73405518a509d38ae84287934bddd8bd2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14873/","number":"14873","commit":"ac47fc5586720077be30c5f09116a36a7d9372fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15139/","number":"15139","commit":"f64cd3fee556435843fd61e55d57794ef841e79d"}]}
PIPELINE-->

alisonelizabeth commented:
> I also found a bug. With the ComboBox component, you can select an option with the enter key. This was also causing the form to submit. I don't think this is desired behavior, especially for fields like alias that have required parameters. I updated the code to prevent this behavior via 9da4267.

After looking into this further, I reverted this workaround. I think it is best addressed in EUI. Opened https://github.com/elastic/eui/issues/2637.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15184/)
* Commit: 35b07e5155cd01fb7ee11bbfb47ba041664eb2dd

### History
* :broken_heart: [Build #15139](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15139/) failed f64cd3fee556435843fd61e55d57794ef841e79d
* :broken_heart: [Build #14873](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14873/) failed ac47fc5586720077be30c5f09116a36a7d9372fe
* :green_heart: [Build #14855](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/) succeeded 266163c73405518a509d38ae84287934bddd8bd2
* :green_heart: [Build #14836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/) succeeded 7423f34b951a068d86c7d00f8ad7bf688c1a6b2b
* :green_heart: [Build #14630](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/) succeeded 0907a5fc4f29136653b6579f5d74ef8ae6a76e80

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14630/","number":"14630","commit":"0907a5fc4f29136653b6579f5d74ef8ae6a76e80"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14836/","number":"14836","commit":"7423f34b951a068d86c7d00f8ad7bf688c1a6b2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14855/","number":"14855","commit":"266163c73405518a509d38ae84287934bddd8bd2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14873/","number":"14873","commit":"ac47fc5586720077be30c5f09116a36a7d9372fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15139/","number":"15139","commit":"f64cd3fee556435843fd61e55d57794ef841e79d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15184/","number":"15184","commit":"35b07e5155cd01fb7ee11bbfb47ba041664eb2dd"}]}
PIPELINE-->
